### DISCUSSION TITLE
CUDA out of memory (SOLVED)

### DISCUSSION THREAD

**Discussion by Elonel5** [hlky/stable-diffusion-webui#671] [Q&A] [answered]
I'm encountering a problem with installing: (followed all the retard instructions like a good retard)

> RuntimeError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 4.00 GiB total capacity; 3.42 GiB already allocated; 0 bytes free; 3.48 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF 

and installation loops
Anything I can do?

**Comment by EliEron** [comment, marked as answer]
The message is telling you that you ran out of VRAM. To run Standard-Diffusion on low VRAM cards you currently have to enable one of the optimized modes. The standard optimized mode lowers VRAM a lot but also massively affects speed. The optimize-turbo mode uses a bit more VRAM but runs with little speed penalty.

To enable the optimized mode open the "relauncher.py" file and change:

```python
optimized = False
```
To:

```Python
optimized = True
```

To enable optimized-turbo do the same thing but for the "optimized_turbo" option. Though only one of them should be enabled (set to True) at a time. 

With only 4GB of VRAM (based on the error message) you will likely have to use the standard optimized mode for now, but there is a PR open right now that will lower VRAM requirements a bit further once merged without a speed penalty. And another that should lower it even further but with a bit of a speed penalty. So you might be able to switch to optimized-turbo in the near future

**Reply by denizbinay** [reply]
I get an simliar error with a colab pro account when i try to generate images in 1024x1024px, even with optimized = true. Is 16gib not enough for 1024x1024?

!!Runtime error (txt2img)!! 
 CUDA out of memory. Tried to allocate 4.00 GiB (GPU 0; 15.90 GiB total capacity; 13.69 GiB already allocated; 1.17 GiB free; 13.89 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
exiting...calling os._exit(0)

**Reply by EliEron** [reply]
Standard-Diffusion is extremely VRAM hungry, especially when you turn up the resolution. Keep in mind that 1024x1024 is actually 4x larger than 512x512 (not 2x as you might assume) since you are doubling both the height and width. 

The optimized mode lowers VRAM usage but it's not a miracle worker. Though as mentioned in my post there have been some recent PRs that have reduced VRAM usage by quite a bit which are not merged yet. I have been experimenting with one of them and it allows me to create 1024x1024 and I don't have close to 16GB of VRAM. Though It's worth noting that since the model was trained to generate 512x512 images it might generates worse images at higher resolutions.

If you are technically inclined you can try out the changes early by merging [this](https://github.com/hlky/stable-diffusion/commit/9d16f26eb4cd645d5cf811be540bb608c5c7a74c.patch) commit into your local copy.

I don't use colab but you should be able to apply the patch by running the following code block:

```bash
!wget https://github.com/hlky/stable-diffusion/commit/9d16f26eb4cd645d5cf811be540bb608c5c7a74c.patch
!git apply 9d16f26eb4cd645d5cf811be540bb608c5c7a74c.patch
```

This assume you are currently in the root directory of the stable-diffusion repo, but I assume most colab scripts place you there after getting the repo.

**Reply by denizbinay** [reply]
it works! very nice. Thank you :-)

**Reply by EliEron** [reply]
Glad to hear it. 

Though to be honest in my testing I have found that generating a 512x512 images an upscaling it via Real-ESRGAN (which is built into the webui), produces images I'm more happy with than the 1024x1024 images natively rendered by the model. Since it seems to duplicate details a lot when rendering at such a high resolution 

Though I would be curious to hear what others experience. It's possible that it's just particularly unsuited for the types of images I'm prompting for. Or that it works better with certain samplers, or the like.

**Reply by TingTingin** [reply]
If you want to prevent duplication you need to describe the objects you don't want duplicated less and the ones you want duplicated more I. E a prompt like **dog** photo is likely to fail as you have one subject **dog** however a simple change like dog photo, in a field of grass, flowers etc will make it where to model is less likely to try to fit every empty space with a dog

**Reply by EliEron** [reply]
I don't have a problem with duplication in general, only when running with resolutions as high as 1024x1024. And I don't think it's down to the prompt, but a limit in the model itself. Since it was designed specifically for 512x512 outputs. 

Here is the output I get when generating your exact prompt in 1024x1024:
![download (3)](https://user-images.githubusercontent.com/22575945/188807887-b22fa147-4b32-4e3e-9176-7927a0189720.png)

And here is the same prompt when ran in 512x512:
![download (11)](https://user-images.githubusercontent.com/22575945/188808320-b86d1b50-d0eb-4c64-ac12-2da99afe850d.png)

And this was a persistent pattern. I ran three generations at 1024x1024 and got three dogs each time, ran three generations at 512x512 and got one dog each time.

And while getting multiple dogs isn't necessarily an issue in images like this, there are image types where the duplicated details make it pretty dull. For instance one of the image types I like to experiment with is Sakura Blossoms. And one thing I noticed was that at 1024x1024 I consistently got results where about 80% of the image consisted of just Sakura blossoms. Like in this image:
![download (18)](https://user-images.githubusercontent.com/22575945/188810063-82b3c8a4-4fb9-4d65-b188-a97c7ab25ef4.png)

Now this image doesn't look bad by itself, but when pretty much every image ends up looking like that it becomes a bit of a problem. In contrast here is a photo with the same prompt in 512x512:
![download (21)](https://user-images.githubusercontent.com/22575945/188813081-d858e3f5-10ce-4191-b0fa-ecc60df00da4.png)

While it can be debated which image looks best, the bottom one is certainly more varied. And that was a consistent pattern as I experimented. I have had some decent success at wider aspect ratios like 704x512 so I'm not saying 512x512 is always best. Just that going to far above the intended 512x512 seems to create repetitive images that aren't quite worth it when AI upscaling looks pretty decent these days. At least for the type of images I have been experimenting with.

**Reply by TingTingin** [reply]
I was saying how to go about trying to get around the model limitation not saying that the model wasn't limited the model was trained on 512x512 images once you go outside that the model will have issues generating images ending up with duplications of subject as the model tries to make sure that the prompt is adhered to in all parts of the image but the aesthetics it was trained on breaks down and you end up with duplication of parts of the image which generally wouldn't happen in good aesthetic images but since the model has never seen those at 1024x1024 it guesses and fails this is a well known issue it seems like you thought I was disagreeing with you which I wasn't you can generally try to get around this doing what I said in the comment above

**Reply by EliEron** [reply]
Ah my mistake then. To me your comment came across as suggesting the issue was not the resolution, but entirely down to a weak prompt. Which is why I tried to illustrate than even with the same prompt the resolution had a decremental effect. It seems we are more or less on the same page then.

Though I personally am not sure you can actually get around this issue even with an ideal prompt. At least not on resolutions that high. As I've personally not found any type of prompt that did not generate a lot of duplication at that resolution.

**Reply by TingTingin** [reply]
The point isn't to prevent duplication the point is to duplicate everything but the thing that you don't want duplicated as duplication is guaranteed to happen you have to pad the prompt with things that when duplicated they wont look as bad the subject say a dog or person will generally look bad when duplicated so you have to reduce the weight of those either burying them with other keywords or moving them down later in the prompt to reduce weight though not too far as you might end up not generating them at all

**Reply by EliEron** [reply]
Ah I got you, that makes more sense. I didn't quite get that from reading your comment the first time, though rereading it I can see where you were coming from. I might give that a try later, as I do usually place the main subject at the front in the prompt.

**Reply by Elonel5** [reply]
It worked! thanks!

**Reply by denizbinay** [reply]
In fact, I experiment a lot with patterns and the results are much better in 1024x1024 than in 512x512. Maybe because there is room for repetition.

**Reply by codefaux** [reply]
I'm receiving a CUDA out of memory error as well, but my situation is a bit different.

A few years back, my partner and I built two identical high-spec gaming machines. The case is different, but the GPU (1080 TI), motherboard, CPU (i7-8600k), RAM (32gb), PSU, SSDs, (you get it) are all identical. My machine runs SD in Linux, his is untested in Linux due to Windows preference. Both machines also run Windows 10 64-bit, same update channel.

My machine works flawlessly with Stable Diffusion in all modes. His crashes during init with a Cuda out-of-memory, before the port is even opened. We've cloned my install to his system, to make sure it wasn't a corrupt download. Fresh install, same steps on both systems. Drivers should match. Ideas?

**Reply by mchaker** [reply]
@codefaux yes, possibly:

Windows could be using some (or more) GPU capacity to render the desktop environment (resource usage would increase when monitor size increases), reducing the available resources for stable-diffusion.

**Comment by mchaker** [comment]
Have you tried the Doggettx optimization? 

Item 1 in this list: https://gist.github.com/mchaker/e0cf180db6849fa03696293b3fc8f485

**Reply by y0himba** [reply]
I'd like to try these, but I am new to git.  How do I implement them in my existing stable-diffusion installation?

**Reply by mchaker** [reply]
I have added instructions to the file. Please let me know if further detail is desired.

**Reply by y0himba** [reply]
Thank you!

**Reply by Elonel5** [reply]
Instructions unclear - how do I install this?
 the link has ... in the middle, what does that index number mean?

**Reply by mchaker** [reply]
Apologies - so, to install, you would copy the changed files from the repository (located at the link) down to your local stable-diffusion folder. I think getting the changed file from the repository is the confusing part?

The index number is for indexing the link in the file, so I can refer to different links by a value better than their "number" in the list. 

What do you feel a better name for it (the index number) would be?

As for the `...`, that is valid in the URL - that is how github denotes "difference between files", I think.

**Reply by Elonel5** [reply]
Thank you for your patience, I don't come from programming and I'm also new to GitHub

I entered _https://github.com/CompVis/stable-diffusion/compare/main...Doggettx:stable-diffusion:main_

How do you I download the new files?
I see the numbers on the side, that look like the index number, but can't find _d8ad218c_

Thank you so much

**Reply by Elonel5** [reply]
I'm actually still lost, can't find d8ad218c, How do I download the new files?

**Reply by mchaker** [reply]
Apologies - so, to download the new files:

1. Go to the link from your post (Doggettx optimizations)
1. Scroll down until you see the file header for `attention.py`:
    ![image](https://user-images.githubusercontent.com/10101544/189238887-280a4fb9-6f0a-4a7a-a482-37279ddf08ae.png)
1. on the right side of that file header, you should see a 3-dots menu
1. click the 3-dots menu and select "view file"
    ![image](https://user-images.githubusercontent.com/10101544/189239026-39e4aa95-13b7-485d-bfa3-2bda3ca283f0.png)
1. On the right side of the page, you should see a "Raw" button. Click that
    ![image](https://user-images.githubusercontent.com/10101544/189239088-9518c21f-2272-4f07-b1b5-ddb092029b96.png)
1. On this page (it should just be a page of code), press Ctrl-S (or Command-S on Mac) to save the page as a file. Save this `attention.py` file under your stable diffusion folder, inside the following path: `(stable-diffusion folder)/ldm/modules/` .
1. Repeat steps 2 through 6 for the next file: `model.py`, placing it in the `(stable-diffusion folder)/ldm/modules/diffusionmodules/` folder.
    ![image](https://user-images.githubusercontent.com/10101544/189239398-5b181039-f4e3-46ed-b852-40244fa18e3f.png)

**Reply by Elonel5** [reply]
Thank you so much KING! Did this, and rendered a 512 x 704 !!!
I appreciate the detailed instructions and could have never figured it out myself!

**Reply by mchaker** [reply]
Haha thanks and awesome!!! Glad to hear it :)
